Added feature to generate Boolean functions from mined event logs.

jrbeaumont opened the pull request:
This feature is designed to work with [Plato](https://github.com/tuura/plato), so a circuit with no known set or reset functions can be simulated to produce event logs, and then mined. The result of this can be stated in a Boolean functiojn format, which can then be converted to asynchronous concepts by Plato. 

@snowleopard This is a first round implementation, please have a look and let me know if there are any issues. All of my testing proves it to work successfully, but if you have any particularly interesting examples you would like to try, let me know :)

For use with Plato, the [boolean](https://github.com/tuura/boolean) repo will also need to be changed, allowing `a+` to represent `a` and `a-` to represent `a'`.

jrbeaumont committed:
Added option for generating Boolean functions from mined logs

jrbeaumont committed:
Added feature to generate Boolean functions from mined event logs

jrbeaumont committed:
Now using stack for building and testing

jrbeaumont commented:
The issue for the CI on this PR is because Travis uses Homebrew, which downloads just the latest versions of GHC and Cabal-install. These do not correctly function with pgminer, which needs `ghc-7.10.3` and `cabal-install-1.22.8`.

snowleopard commented:
> These do not correctly function with pgminer, which needs `ghc-7.10.3`

Why doesn't it work with the latest GHC? It would be nice to fix the incompatibility, I'm sure it's not difficult.

jrbeaumont commented:
Homebrew forces it to use `ghc-8.2.2`. Is that OK?

snowleopard commented:
I think it would be nice if you make sure the code compiles with all recent GHC versions, if possible.

jrbeaumont commented:
OK I'll sort that.

jrbeaumont committed:
Set range for extra to be safe

jrbeaumont commented:
@snowleopard: The repo has been switched to stack, and all is working correctly :)

snowleopard commented on pgminer/Tuura/PGminer/Main.hs:
```diff
@@ -35,7 +36,9 @@ main = do
         co a b = cache Map.! (min a b, max a b)
         graphs = map (fmap decode) $ reduceLog co log
-        result = unlines . addIds $ map printGraphExpr graphs
+        result = if (optBool options /= "")
```
Use pattern matching `case (optBool options)` here instead of `/= ""`, it usually leads to cleaner code.

snowleopard commented on pgminer/Tuura/PGminer/Options.hs:
```diff
@@ -32,6 +34,9 @@ options =
       "Split traces with multiple event occurrences"
+    , Option ['b'] ["bool"]
+      (ReqArg (\f opts -> return opts { optBool = f}) "STRING")
+      "Output variable"
```
`Output variable` is not a very clear hint for the `-b` flag.

snowleopard commented on src/Tuura/Function.hs:
```diff
@@ -0,0 +1,47 @@
+
+-- data Function = Open String | Closed String deriving (Eq, Ord)
+
+instance Graph Function where
```
I just realised that we should probably switch to using `alga` library for working with graphs. But that's future work.

jrbeaumont replied:
I agree actually, it would make some stuff really nice and neat :)

snowleopard commented on src/Tuura/Function.hs:
```diff
@@ -0,0 +1,47 @@
+foldFunction e v o c s ge = case ge of
+        Empty       -> e
+        Vertex x    -> v x
+        Overlay x y -> if test s x
```
Can you try to rewrite this using guards `| test s x = ...` instead of `if` statements? It might look nicer.

snowleopard replied:
I'm pretty sure this whole thing can be simplified a lot using the Alga library.

jrbeaumont replied:
Guards won't work inside a case statement. But I can use further case statements instead. This may or may not add to confusion, but wait for the next commit to view, and we can decide.

snowleopard commented on src/Tuura/Function.hs:
```diff
@@ -0,0 +1,47 @@
+printFunction (Open   s) = s
+printFunction (Closed s) = s
+
+test :: String -> GraphExpr String -> Bool
```
This, for example, is nothing else as `hasVertex`: 

http://hackage.haskell.org/package/algebraic-graphs-0.0.5/docs/Algebra-Graph.html#v:hasVertex

snowleopard commented:
@jrbeaumont Thanks! I've left a couple of comments.

Can you also update the documentation, including a simple example?

jrbeaumont committed:
Tidied up some code

jrbeaumont committed:
Added a clearer message for the -b option

jrbeaumont committed:
Use case statements instead of if statements for neater code

jrbeaumont committed:
Changed info after move to Stack, added example for -b option

jrbeaumont commented:
@snowleopard I have made those changes. To see the changes to the README, please go to [my forked repo](https://github.com/jrbeaumont/process-mining), so you can see the changes in action :)

jrbeaumont commented:
Oh I'm also going to edit `tuura/boolean` too.

snowleopard commented on README.md:
````diff
@@ -85,20 +104,65 @@ p2 = a -> b + b -> c + c -> e
+This will produce the following function:
+
+```
+a+ * b+ | a+ | b+
````
Why do you use `*` instead of more common `&` to denote the AND operator?

jrbeaumont committed:
AND operation now represented by & instead.

jrbeaumont committed:
Signal transitions will now remove `+` and replace `-` with `'`

jrbeaumont commented:
@snowleopard `+` is now removed at the end of variables, and `-` at the end of variables is replaced with `'`. I have also updated the README to reflect this. `&` is also now used for AND instead of `*`.

snowleopard commented on src/Tuura/Function.hs:
```diff
@@ -3,17 +3,22 @@ module Tuura.Function (
     empty       = Closed ""
-    vertex      = Closed . id
-    overlay p q = Open $ printFunction p ++ " * " ++ printFunction q
+    vertex v    = Closed . var $ id v
```
Well, since `id v = v` by the definition of `id`, we can simply use `v` here :)

snowleopard commented on src/Tuura/Function.hs:
```diff
@@ -3,17 +3,22 @@ module Tuura.Function (
+var :: [Char] -> [Char]
+var v
+  | "+" `isSuffixOf` v = init v
+  | "-" `isSuffixOf` v = (init v) ++ "'"
```
Don't need parentheses around `init v` here.

snowleopard commented:
@jrbeaumont Thanks, looks good! Just a couple more comments and I'll merge this.

jrbeaumont committed:
Simplified some code

jrbeaumont commented:
@snowleopard Thanks for the tips. Fixed those issues!

snowleopard commented:
@jrbeaumont Thanks! Merged.
